clang/lib/CodeGen/BackendUtil.cpp

```diff
@@ -1035,31 +1033,33 @@ void EmitAssemblyHelper::RunOptimizationPipeline(
       MPM.addPass(createModuleToFunctionPassAdaptor(MemProfilerPass()));
       MPM.addPass(ModuleMemProfilerPass());
     }
-  }
-  if (LangOpts.SYCLIsDevice) {
-    MPM.addPass(SYCLMutatePrintfAddrspacePass());
-    if (!CodeGenOpts.DisableLLVMPasses && LangOpts.EnableDAEInSpirKernels)
-      MPM.addPass(DeadArgumentEliminationSYCLPass());
-  }
 
-  // Add SPIRITTAnnotations pass to the pass manager if
-  // -fsycl-instrument-device-code option was passed. This option can be used
-  // only with spir triple.
-  if (LangOpts.SYCLIsDevice && CodeGenOpts.SPIRITTAnnotations) {
-    assert(TargetTriple.isSPIR() &&
-           "ITT annotations can only be added to a module with spir target");
-    MPM.addPass(SPIRITTAnnotationsPass());
-  }
+    if (LangOpts.SYCLIsDevice) {
+      MPM.addPass(SYCLMutatePrintfAddrspacePass());
+      if (LangOpts.EnableDAEInSpirKernels)
+        MPM.addPass(DeadArgumentEliminationSYCLPass());
+    }
+
+    // Add SPIRITTAnnotations pass to the pass manager if
+    // -fsycl-instrument-device-code option was passed. This option can be used
+    // only with spir triple.
+    if (LangOpts.SYCLIsDevice && CodeGenOpts.SPIRITTAnnotations) {
+      assert(TargetTriple.isSPIR() &&
+             "ITT annotations can only be added to a module with spir target");
+      MPM.addPass(SPIRITTAnnotationsPass());
+    }
 
-  // Allocate static local memory in SYCL kernel scope for each allocation
-  // call. It should be called after inlining pass.
-  if (LangOpts.SYCLIsDevice) {
-    // Group local memory pass depends on inlining. Turn it on even in case if
-    // all llvm passes or SYCL early optimizations are disabled.
-    // FIXME: Remove this workaround when dependency on inlining is eliminated.
-    if (CodeGenOpts.DisableLLVMPasses)
-      MPM.addPass(AlwaysInlinerPass(false));
-    MPM.addPass(SYCLLowerWGLocalMemoryPass());
+    // Allocate static local memory in SYCL kernel scope for each allocation
+    // call. It should be called after inlining pass.
+    if (LangOpts.SYCLIsDevice) {
+      // Group local memory pass depends on inlining. Turn it on even in case if
+      // all llvm passes or SYCL early optimizations are disabled.
+      // FIXME: Remove this workaround when dependency on inlining is
+      // eliminated.
+      if (CodeGenOpts.DisableSYCLEarlyOpts)
```

Comment:
Can this be if (CodeGenOpts.OptimizationLevel == 0). If so, I don't understand why we need this flag

Reply:
The if should be removed completely, together with addition of `AlwaysInliner` pass. The situation when `SYCLLowerWGLocalMemory` pass is enabled, but inliner isn't enabled is not reachable anymore